internal/command/group_users.go

```diff
@@ -0,0 +1,167 @@
+package command
+
+import (
+	"context"
+	"slices"
+	"time"
+
+	"github.com/zitadel/logging"
+
+	"github.com/zitadel/zitadel/internal/domain"
+	repo "github.com/zitadel/zitadel/internal/repository/group"
+	"github.com/zitadel/zitadel/internal/telemetry/tracing"
+	"github.com/zitadel/zitadel/internal/zerrors"
+)
+
+type AddUsersToGroupResponse struct {
+	*domain.ObjectDetails
+	FailedUserIDs []string
+}
+
+func (c *Commands) AddUsersToGroup(ctx context.Context, groupID string, userIDs []string) (_ *AddUsersToGroupResponse, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.EndWithError(err) }()
+
+	// precondition: check whether the group exists
+	group, err := c.getGroupWriteModelByID(ctx, groupID, "")
+	if err != nil {
+		return nil, err
+	}
+	if !group.State.Exists() {
+		return nil, zerrors.ThrowPreconditionFailed(nil, "CMDGRP-eQfeur", "Errors.Group.NotFound")
+	}
+
+	// check whether the requester has permissions to add users to the group
+	err = c.checkPermissionAddUserToGroup(ctx, group.ResourceOwner, group.AggregateID)
+	if err != nil {
+		return nil, err
+	}
+
+	// add the users to the group
+	details, failedUserIDs, err := c.addUsersToGroup(ctx, group.ResourceOwner, group.AggregateID, userIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AddUsersToGroupResponse{
+		FailedUserIDs: failedUserIDs,
+		ObjectDetails: details,
+	}, nil
+}
+
+func (c *Commands) RemoveUsersFromGroup(ctx context.Context, groupID string, userIDs []string) (_ *domain.ObjectDetails, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.EndWithError(err) }()
+
+	// precondition: check whether the group exists
+	group, err := c.getGroupWriteModelByID(ctx, groupID, "")
+	if err != nil {
+		return nil, err
+	}
+	if !group.State.Exists() {
+		return nil, zerrors.ThrowPreconditionFailed(nil, "CMDGRP-JRBnLw", "Errors.Group.NotFound")
+	}
+
+	// check whether the requester has permissions to remove users from the group
+	err = c.checkPermissionRemoveUserFromGroup(ctx, group.ResourceOwner, group.AggregateID)
+	if err != nil {
+		return nil, err
+	}
+
+	// remove duplicate userIDs
+	uniqueUserIDs := removeDuplicateUserIDs(userIDs)
+
+	groupUsersWriteModel, err := c.getGroupUsersWriteModel(ctx, group.ResourceOwner, groupID, uniqueUserIDs)
```

Comment:
`getGroupUsersWriteModel` and `getGroupWriteModelByID` can be combined, so you only use 1 query to the DB.